pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/pipeline/PipelineBreakerOperator.java

```diff
@@ -62,70 +55,55 @@ public Map<Integer, List<TransferableBlock>> getResultMap() {
-        _finalBlock = TransferableBlockUtils.getErrorTransferableBlock(QueryException.EXECUTION_TIMEOUT_ERROR);
-        constructErrorResponse(_finalBlock);
-        return _finalBlock;
-      }
-
-      Map.Entry<Integer, Operator<TransferableBlock>> worker = _workerEntries.getLast();
-      TransferableBlock block = worker.getValue().nextBlock();
-
-      if (block == null) {
-        if (LOGGER.isDebugEnabled()) {
-          LOGGER.debug("==[PB]== Null block on " + _context.getId() + " worker " + worker.getKey());
+    if (_errorBlock != null) {
+      return _errorBlock;
+    }
+    // NOTE: Put an empty list for each worker in case there is no data block returned from that worker
+    if (_workerMap.size() == 1) {
+      Map.Entry<Integer, Operator<TransferableBlock>> entry = _workerMap.entrySet().iterator().next();
+      List<TransferableBlock> dataBlocks = new ArrayList<>();
+      _resultMap = Collections.singletonMap(entry.getKey(), dataBlocks);
+      Operator<TransferableBlock> operator = entry.getValue();
+      TransferableBlock block = operator.nextBlock();
+      while (!block.isSuccessfulEndOfStreamBlock()) {
+        if (block.isErrorBlock()) {
+          _errorBlock = block;
+          return block;
         }
-        continue;
+        dataBlocks.add(block);
+        block = operator.nextBlock();
       }
-
-      // Release the mailbox worker when the block is end-of-stream
-      if (block.isSuccessfulEndOfStreamBlock()) {
-        _workerEntries.removeLast();
-        continue;
+    } else {
+      _resultMap = new HashMap<>();
+      for (int workerKey : _workerMap.keySet()) {
+        _resultMap.put(workerKey, new ArrayList<>());
       }
-
-      if (block.isErrorBlock()) {
-        _finalBlock = block;
-      }
-      if (LOGGER.isTraceEnabled()) {
-        LOGGER.trace("==[PB]== Returned block from : " + _context.getId() + " block: " + block);
-      }
-      _resultMap.get(worker.getKey()).add(block);
-      return block;
-    }
-    if (System.currentTimeMillis() > _context.getDeadlineMs()) {
-      _finalBlock = TransferableBlockUtils.getErrorTransferableBlock(QueryException.EXECUTION_TIMEOUT_ERROR);
-      return _finalBlock;
-    } else if (_finalBlock == null) {
-      if (LOGGER.isDebugEnabled()) {
-        LOGGER.debug("==[PB]== Finished : " + _context.getId());
+      // Keep polling from every operator in round-robin fashion
+      Queue<Map.Entry<Integer, Operator<TransferableBlock>>> entries = new ArrayDeque<>(_workerMap.entrySet());
+      while (!entries.isEmpty()) {
+        Map.Entry<Integer, Operator<TransferableBlock>> entry = entries.poll();
+        TransferableBlock block = entry.getValue().nextBlock();
+        if (block.isErrorBlock()) {
+          _errorBlock = block;
+          return block;
+        }
+        if (block.isDataBlock()) {
+          _resultMap.get(entry.getKey()).add(block);
+          entries.offer(entry);
+        }
```

Comment:
Do we check for statsMap of endOfStream data block?

Reply:
It is tracked within the op-chain